### PR TITLE
Refactor SSE Matmul to align with AVX codegen, added I8 and BF16 support

### PR DESCRIPTION
**Pull request by alheinecke**

### CODE REVIEW

**Comment by alheinecke** [comment]
This is for SGEMM a performance sweep over 65K GEMMs in the 1-1-1 : 100:100:100 window:
image

**Comment by alheinecke** [comment]
@egeor : this is still in draft as the int8 stuff has not yet been added, but please review if all your GEMM changes have been merged correctly. I plan to merge this one later this week.

File: src/generator_common_x86.c [hunk 1]
```diff
@@ -135,10 +135,28 @@ void libxsmm_x86_instruction_unified_vec_move( libxsmm_generated_code* io_genera
 
   if (io_generated_code->arch < LIBXSMM_X86_AVX) {
     if (i_use_masking > 0) {
-      LIBXSMM_HANDLE_ERROR( io_generated_code, LIBXSMM_ERR_GENERAL );
-      return;
+      if ( (i_vmove_instr == LIBXSMM_X86_INSTR_MOVUPS) ||
+           (i_vmove_instr == LIBXSMM_X86_INSTR_MOVUPS)    ) {
```
Comment (egeor) [inline src/generator_common_x86.c:139]: I guess meant to be MOVAPS?

File: src/generator_common_x86.c [hunk 2]
```diff
@@ -1378,6 +1396,44 @@ void libxsmm_generator_prepare_coeffs_tanh_ps_rational_78_avx( libxsmm_generated
...
+  float c3_array[4] = { 36.0f, 36.0f, 36.0f, 36.0f };
+  float c1_d_array[4] = { 945945.0f, 945945.0f, 945945.0f, 945945.0f};
+  float c2_d_array[4] = { 51975.0f, 51975.0f, 51975.0f, 51975.0f };
+  float c3_d_array[4] = { 630.0f, 630.0f, 630.0f, 630.0f };
+  float hi_b_array[4] = { 4.97f, 4.97f, 4.97f, 4.97f };
+  float lo_b_array[4] = { -4.97f, -4.97f, -4.97f, -4.97f };
+  float ones_array[4] = { 1.0f, 1.0f, 1.0f, 1.0f };
+  float neg_ones_array[4] = { -1.0f, -1.0f, -1.0f, -1.0f };
+
+  libxsmm_x86_instruction_full_vec_load_of_constants ( io_generated_code, (const unsigned char *) c0_array, "c0_array_", 'y', i_vec_c0);
```
Comment (egeor) [inline src/generator_common_x86.c:1424]: Shouldn't these be xmm loads for the constants?

File: src/generator_common_x86.c [hunk 3]
```diff
@@ -3274,3 +3489,424 @@ void libxsmm_generator_initialize_avx_mask( libxsmm_generated_code* io_generated
...
+  int l_displacement = 0;
+  unsigned int l_mask_count = i_mask_count;
+
+  LIBXSMM_UNUSED( i_save_gp_reg_tmp_inside );
+
+  /* allocate 1/4 cache lines on the stack, TODO: make sure taht the stack pointer is 64 bytealigned for perf */
+  libxsmm_x86_instruction_alu_imm( io_generated_code, LIBXSMM_X86_INSTR_SUBQ, LIBXSMM_X86_GP_REG_RSP, 16 );
+
+  /* write 0 to this cache line */
+  libxsmm_x86_instruction_vec_compute_3reg( io_generated_code, LIBXSMM_X86_INSTR_VPXORD, 'y',
```
Comment (egeor) [inline src/generator_common_x86.c:3511]: xmm vname?

File: src/generator_common_x86.c [hunk 4]
```diff
@@ -3274,3 +3489,424 @@ void libxsmm_generator_initialize_avx_mask( libxsmm_generated_code* io_generated
...
+  int l_displacement = 0;
+  unsigned int l_mask_count = i_mask_count;
+
+  LIBXSMM_UNUSED( i_save_gp_reg_tmp_inside );
+
+  /* allocate 1/4 cache lines on the stack, TODO: make sure taht the stack pointer is 64 bytealigned for perf */
+  libxsmm_x86_instruction_alu_imm( io_generated_code, LIBXSMM_X86_INSTR_SUBQ, LIBXSMM_X86_GP_REG_RSP, 16 );
+
+  /* write register into this cache line */
+  libxsmm_x86_instruction_vec_move( io_generated_code, io_generated_code->arch, LIBXSMM_X86_INSTR_VMOVUPS, LIBXSMM_X86_GP_REG_RSP, LIBXSMM_X86_GP_REG_UNDEF, 0,  0, 'y', i_vec_reg_in, 0, 0, 1 );
```
Comment (egeor) [inline src/generator_common_x86.c:3573]: xmm vname?

File: src/generator_common_x86.c [hunk 5]
```diff
@@ -3274,3 +3489,424 @@ void libxsmm_generator_initialize_avx_mask( libxsmm_generated_code* io_generated
...
+  int l_displacement = 0;
+  unsigned int l_mask_count = i_mask_count;
+
+  LIBXSMM_UNUSED( i_save_gp_reg_tmp_inside );
+
+  /* allocate 1/4 cache lines on the stack, TODO: make sure taht the stack pointer is 64 bytealigned for perf */
+  libxsmm_x86_instruction_alu_imm( io_generated_code, LIBXSMM_X86_INSTR_SUBQ, LIBXSMM_X86_GP_REG_RSP, 16 );
+
+  /* write 0 to this cache line */
+  libxsmm_x86_instruction_vec_compute_3reg( io_generated_code, LIBXSMM_X86_INSTR_VPXORD, 'y',
```
Comment (egeor) [inline src/generator_common_x86.c:3629]: xmm vname?

File: src/generator_common_x86.c [hunk 6]
```diff
@@ -3274,3 +3489,424 @@ void libxsmm_generator_initialize_avx_mask( libxsmm_generated_code* io_generated
...
+  }
+  if ( l_mask_count == 1 ) {
+    libxsmm_x86_instruction_alu_mem( io_generated_code, LIBXSMM_X86_INSTR_MOVW, i_gp_reg_base, i_reg_idx, i_scale, i_displacement+l_displacement, i_gp_reg_tmp, 0);
+    libxsmm_x86_instruction_alu_mem( io_generated_code, LIBXSMM_X86_INSTR_MOVW, LIBXSMM_X86_GP_REG_RSP, LIBXSMM_X86_GP_REG_UNDEF, 0, l_displacement, i_gp_reg_tmp, 1);
+    l_displacement += 2;
+    l_mask_count -= 1;
+  }
+
+  /* load the vector register */
+  libxsmm_x86_instruction_vec_move( io_generated_code, io_generated_code->arch, LIBXSMM_X86_INSTR_VMOVUPS, LIBXSMM_X86_GP_REG_RSP, LIBXSMM_X86_GP_REG_UNDEF, 0,  0, 'y', i_vec_reg_out, 0, 0, 0 );
```
Comment (egeor) [inline src/generator_common_x86.c:3660]: cmm vname?

**Review by egeor — COMMENTED** [review]
LGTM. Wrt keeping track of m advance for the relu sse bitmask logistics -- this is I am afraid the way to go for now... I left some comments wet wrong vnames used in some cases for SSE; I guess they didn't show up as Illegal instructions because you were testing on an AVX machine?

File: src/generator_common_x86.c (hunk 1 above)
Comment (alheinecke) [reply src/generator_common_x86.c:139]: yepp

File: src/generator_common_x86.c (hunk 3 above)
Comment (alheinecke) [reply src/generator_common_x86.c:3511]: yepp and also 2 operands... all load/store masked 16 and 8 bit have not been tested / implemented yet. They are just C&P from AVX.

File: src/generator_common_x86.c (hunk 5 above)
Comment (alheinecke) [reply src/generator_common_x86.c:3629]: see above

File: src/generator_common_x86.c [hunk 7]
```diff
@@ -3274,3 +3489,424 @@ void libxsmm_generator_initialize_avx_mask( libxsmm_generated_code* io_generated
...
+  int l_displacement = 0;
+  unsigned int l_mask_count = i_mask_count;
+
+  LIBXSMM_UNUSED( i_save_gp_reg_tmp_inside );
+
+  /* allocate 1/4 cache lines on the stack, TODO: make sure taht the stack pointer is 64 bytealigned for perf */
+  libxsmm_x86_instruction_alu_imm( io_generated_code, LIBXSMM_X86_INSTR_SUBQ, LIBXSMM_X86_GP_REG_RSP, 16 );
+
+  /* write register into this cache line */
+  libxsmm_x86_instruction_vec_move( io_generated_code, io_generated_code->arch, LIBXSMM_X86_INSTR_VMOVUPS, LIBXSMM_X86_GP_REG_RSP, LIBXSMM_X86_GP_REG_UNDEF, 0,  0, 'y', i_vec_reg_in, 0, 0, 1 );
```
[the remark this answers is not in the document]
Comment (alheinecke) [reply src/generator_common_x86.c:3685]: see above

File: src/generator_common_x86.c (hunk 6 above)
Comment (alheinecke) [reply src/generator_common_x86.c:3660]: see above

File: src/generator_common_x86.c (hunk 2 above)
Comment (alheinecke) [reply src/generator_common_x86.c:1424]: fixed

**Pull request merged**